python/ray/serve/_private/http_state.py

```diff
@@ -72,65 +76,137 @@ def set_status(self, status: HTTPProxyStatus) -> None:
         """Updates _actor_details with passed in kwargs."""
         details_kwargs = self._actor_details.dict()
         details_kwargs.update(kwargs)
         self._actor_details = HTTPProxyDetails(**details_kwargs)
 
     def update(self):
+        """Update the status of the current HTTP proxy.
+
+        1) When the HTTP proxy is already shutting down, do nothing.
+        2) When the HTTP proxy is starting, call ready object. If ready object returns a
+        successful call, set status to HEALTHY. If the ready call has any exception or
+        timeout, count towards 1 of the consecutive health check failures and retry on
+        the next update call. The status is only set to UNHEALTHY when all retries have
+        exhausted.
+        3) When the HTTP proxy already has an in-progress health check. If health check
+        object returns a successful call, set status to HEALTHY. If the call has any
+        exception or timeout, count towards 1 of the consecutive health check failures
+        and retry on the next update call. The status is only set to UNHEALTHY when all
+        retries have exhausted.
+        4) When the HTTP proxy need to setup another health check (when none of the
+        above met and the time since the last health check is longer than
+        PROXY_HEALTH_CHECK_PERIOD_S with some margin). Reset
+        self._last_health_check_time and set up a new health check object so the next
+        update can call healthy check again.
+        """
         if self._shutting_down:
             return
 
         if self._status == HTTPProxyStatus.STARTING:
             finished, _ = ray.wait([self._ready_obj_ref], timeout=0)
             if finished:
-                self._ready_obj_ref = None
                 try:
                     worker_id, log_file_path = json.loads(ray.get(finished[0]))
-                    self.set_status(HTTPProxyStatus.HEALTHY)
+                    self.try_update_status(HTTPProxyStatus.HEALTHY)
                     self.update_actor_details(
                         worker_id=worker_id,
                         log_file_path=log_file_path,
                         status=self._status,
                     )
                 except Exception:
-                    self.set_status(HTTPProxyStatus.UNHEALTHY)
+                    self.try_update_status(HTTPProxyStatus.UNHEALTHY)
                     logger.warning(
-                        "Unexpected error occured when checking readiness of HTTP "
+                        "Unexpected error occurred when checking readiness of HTTP "
                         f"Proxy on node {self._node_id}:\n{traceback.format_exc()}"
                     )
+            elif (
+                time.time() - self._last_health_check_time
+                > DEFAULT_READY_CHECK_TIMEOUT_S
+            ):
+                # Ready check hasn't returned and the timeout is up, consider it failed.
+                self.try_update_status(HTTPProxyStatus.UNHEALTHY)
```

Comment:
Will this just loop for 3 times in quick succession after we first pass the `DEFAULT_HEALTH_CHECK_TIMEOUT_S` threshold? If so should we just set the status directly instead of waiting for 3 more iterations?

Reply:
Oh I see, yea will just use `set_status()` here